### PR TITLE
[feature] extend PCU driver to handle several PCUs connected

### PR DESCRIPTION
**Pull request by pieleric**

### CODE REVIEW

File: src/odemis/driver/powerctrl.py [hunk 1]
```diff
@@ -44,46 +46,62 @@ class PowerControlUnit(model.PowerSupplier):
...
         port (str): port name
         pin_map (dict of str -> int): names of the components
           and the pin where the component is connected.
         delay (dict str -> float): time to wait for each component after it is
             turned on.
         init (dict str -> boolean): turn on/off the corresponding component upon
             initialization.
-        ids (list str): EEPROM ids expected to be detected during initialization.
+        ids (None or list[str]): EEPROM ids expected to be detected during initialization.
+        All the IDs must be found. If not, another device is search for.
```
Comment (tepals) [inline src/odemis/driver/powerctrl.py:62]: what do you mean with "another device is search for"?

File: src/odemis/driver/powerctrl.py [hunk 2]
```diff
@@ -44,46 +46,62 @@ class PowerControlUnit(model.PowerSupplier):
...
+            self._powered = []
 
         if not check_power and ids:
             raise ValueError(f"check_power can only be disabled if no ids are specified, but got {ids}")
 
         self._check_power = check_power
 
-        # TODO: catch errors and convert to HwError
+        required_ids = ids or []
+        if not all(isinstance(eid, str) and len(eid) >= 8 for eid in required_ids):
```
Comment (tepals) [inline src/odemis/driver/powerctrl.py:90]:
```suggestion
        if not all(isinstance(rid, str) and len(rid) >= 8 for rid in required_ids):
```

File: src/odemis/driver/powerctrl.py [hunk 3]
```diff
@@ -44,46 +46,62 @@ class PowerControlUnit(model.PowerSupplier):
...
+            raise ValueError(f"ids must be a list of strings (of at least 8 chars), but got '{ids}'")
+
+        excluded_ids = excluded_ids or []
+        if not all(isinstance(eid, str) and len(eid) >= 8 for eid in excluded_ids):
+            raise ValueError(f"excluded_ids must be a list of strings (of at least 8 chars), but got '{excluded_ids}'")
```
Comment (tepals) [inline src/odemis/driver/powerctrl.py:95]: Does it make sense to mention in the docstring that the ids and excluded_ids must be at least 8 chars?

File: src/odemis/driver/powerctrl.py [hunk 4]
```diff
@@ -44,46 +46,62 @@ class PowerControlUnit(model.PowerSupplier):
     communication with the PCU firmware.
     '''
 
-    def __init__(self, name, role, port, pin_map=None, delay=None, init=None, ids=None,
+    def __init__(self, name, role, port, pin_map=None, delay=None, init=None,
+                 ids: Optional[List[str]] = None,
+                 excluded_ids: Optional[List[str]] = None,
                  termination=None, check_power=True, **kwargs):
-        '''
+        """
```
Comment (tepals) [inline src/odemis/driver/powerctrl.py:53]: can you update this docstring to match the guidelines? e.g. `:param port:`

File: src/odemis/driver/powerctrl.py [hunk 5]
```diff
@@ -380,19 +410,43 @@ def _findDevice(self, ports):
...
+            # (only if EEPROM IDs are requested as this can cause failures on some devices)
+            if required_ids or excluded_ids:
+                try:
+                    mem_ids = self._getIdentities()
+                except HwError:
+                    logging.info("Skipping port %s as EEPROMs cannot be read", n)
+                    error_no_power = True
+                    continue
+                logging.info("PCU on port %s is connected to EEPROMs: %s", n, mem_ids)
+                if not all(eid in mem_ids for eid in required_ids):
```
Comment (tepals) [inline src/odemis/driver/powerctrl.py:433]:
```suggestion
                if not all(rid in mem_ids for rid in required_ids):
```

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/driver/powerctrl.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/driver/powerctrl.py:62]: Search for another PCU. I've updated the docstring to clarify this text.

File: src/odemis/driver/powerctrl.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/driver/powerctrl.py:95]: All EEPROM IDs are above 8 chars. So it'd be redundant. I've just added this check to detect typos in the arguments.

File: install/linux/usr/bin/odemis-mic-selector [hunk 6]
```diff
@@ -64,43 +63,70 @@ def count_usb_device(usbid):
...
+def find_backend_pcus() -> List[model.HwComponent]:
+    """
+    return (list of HwComponent): all the power control units found
+    :raises: IOError if no power control unit is found
+    """
+    pcus = []
+    try:
+        comps = model.getComponents()
+    except (Pyro4.errors.CommunicationError, IOError, LookupError):
+        raise IOError("Failed to access the backend")
```
Comment (sneepstefan) [inline install/linux/usr/bin/odemis-mic-selector:75]: Wouldn't it be nicer to split this up and make a distinction between backend not accessible and backend active but no PCU found?
```suggestion
    except (Pyro4.errors.CommunicationError, IOError):
        raise IOError("Failed to access the backend")
    except LookupError:
        raise LookupError("Failed to find any PCU")    
```

File: install/linux/usr/bin/odemis-mic-selector [hunk 7]
```diff
@@ -64,43 +63,70 @@ def count_usb_device(usbid):
...
     terminate = False
     try:
-        pcu = model.getComponent(role="power-control")
+        pcus = find_backend_pcus()
         logging.debug("Using the backend to access the power controller")
-    except (Pyro4.errors.CommunicationError, IOError, LookupError):
-        logging.debug("Failed to access the backend, will try directly")
-        pcu = CLASS_PCU(**KWARGS_PCU)
+    except IOError:
+        pcus = connect_all_pcus()
```
Comment (sneepstefan) [inline install/linux/usr/bin/odemis-mic-selector:108]: This part is quite confusing, if the backend cannot be accessed and if no PCU is found, try to connect all pcu's?
Maybe add some more comments to clarify this way of connecting?

**Review by sneepstefan — CHANGES_REQUESTED** [review]

File: install/linux/usr/bin/odemis-mic-selector (hunk 6 above)
Comment (pieleric) [reply install/linux/usr/bin/odemis-mic-selector:75]: All these exceptions actually are a sign that there is no back-end. If there is a backend, but no PCU, that's later... and normally it's fine to return an empty list.
The main issue is the docstring which states that it returns IOError if there are no PCU, actually, it's if there is no backend. -> fixed.

**Pull request merged**